lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3461,11 +3461,26 @@ void replaceScalarArrayGEPWithVectorArrayGEP(User *GEP, Value *VectorArray,
   Value *CompIdx = Builder.CreateAnd(ArrayIdx, mask);
   Value *NewGEP = Builder.CreateGEP(
       VecPtr, {ConstantInt::get(CompIdx->getType(), 0), CompIdx});
-  GEP->replaceAllUsesWith(NewGEP);
+
+  if (isa<ConstantExpr>(GEP) && isa<Instruction>(NewGEP)) {
+    if (!ReplaceConstantWithInst(cast<Constant>(GEP), NewGEP, Builder)) {
+      // If new instructions unable to be used, clean them up.
+      if (NewGEP->user_empty())
+        cast<Instruction>(NewGEP)->eraseFromParent();
+      if (VecPtr->user_empty())
+        cast<Instruction>(VecPtr)->eraseFromParent();
```

Comment:
Those will never be instructions (always constant expressions).  This is because we are replacing a GEP that used to be a constant expression using a constant pointer with one that needs to be an instruction because the new pointer is an instruction instead.  That means when we create the math ops, we actually just get constant expression evaluation with nothing we should worry about cleaning up at this point.

Reply:
Oh, hang on, that's not right.  They may not be constant expressions, I was thinking of the replace-constant-with-instruction path for some reason.

Ok, we can delete these if they are unused instructions, though they are less consequential due to not impacting uses of VectorArray and the ability to clean up other intermediates related to that, which could have other impacts before we get to DCE.